src/wms/src/main/java/org/geoserver/wms/clip/ClipWMSGetMapCallBack.java

```diff
@@ -0,0 +1,176 @@
+
+    private static final WKTReader2 reader = new WKTReader2();
+
+    private static final Pattern SRID_REGEX = Pattern.compile("SRID=[0-9].*");
+
+    @Override
+    public GetMapRequest initRequest(GetMapRequest request) {
+        return request;
+    }
+
+    @Override
+    public void initMapContent(WMSMapContent mapContent) {}
+
+    @Override
+    public Layer beforeLayer(WMSMapContent mapContent, Layer layer) {
+
+        // find and parse WKT geometry from WMS request
+        //
+        Geometry wktGeom = getClipGeometry(mapContent.getRequest());
+        if (wktGeom == null) return layer;
+
+        Geometry bboxGeom = JTS.toGeometry(mapContent.getRequest().getBbox());
+        // check: if wkt area fully contains bbox
+        if (wktGeom.covers(bboxGeom)) return layer;
+        try {
+            if (layer instanceof FeatureLayer) {
+
+                // wrap around
+                FeatureLayer fl = (FeatureLayer) layer;
+
+                ClippedFeatureSource clippedFS =
+                        new ClippedFeatureSource(layer.getFeatureSource(), wktGeom);
+                FeatureLayer clippedLayer =
+                        new FeatureLayer(clippedFS, fl.getStyle(), fl.getTitle());
+                BeanUtilsBean2.getInstance().copyProperties(clippedLayer, fl);
+                fl.getUserData().putAll(layer.getUserData());
+                return clippedLayer;
+
+            } else if (layer instanceof GridReaderLayer) {
+
+                GridReaderLayer gr = (GridReaderLayer) layer;
+                // wrap
+                CroppedGridCoverage2DReader croppedGridReader =
+                        new CroppedGridCoverage2DReader(gr.getReader(), wktGeom);
+                GridReaderLayer croppedGridLayer =
+                        new GridReaderLayer(croppedGridReader, layer.getStyle());
+                BeanUtilsBean2.getInstance().copyProperties(croppedGridLayer, gr);
+                croppedGridLayer.getUserData().putAll(layer.getUserData());
+                return croppedGridLayer;
+            }
+        } catch (Exception e) {
+            LOGGER.severe("Error occurred while clipping layer " + layer.getTitle());
+            LOGGER.log(Level.SEVERE, e.getMessage(), e);
+            return layer;
+        }
+
+        return layer;
+    }
+
+    @Override
+    public WMSMapContent beforeRender(WMSMapContent mapContent) {
+        return mapContent;
+    }
+
+    @Override
+    public WebMap finished(WebMap map) {
+        return map;
+    }
+
+    @Override
+    public void failed(Throwable t) {}
+
+    private static synchronized Geometry readGeometry(
+            final String wkt, final CoordinateReferenceSystem mapCRS) throws Exception {
+        String[] wktContents = wkt.split(";");
+        Geometry geom = reader.read(wktContents[wktContents.length - 1]);
+        // geom.setSRID(CRS.lookupEpsgCode(mapCRS, false));
+        if (!(geom.getClass().isAssignableFrom(Polygon.class)

```

Comment:
Throw a service exception in this case, indicating that the content of "clip" is invalid.
Something like (untested): ``throw new ServiceException("Clip must be a polygon or multipolygon", "InvalidParameterValue", "clip")``